### DISCUSSION TITLE
[PROPOSAL]: refactor zero/gemini folder structure

### DISCUSSION THREAD

**Discussion by ver217** [hpcaitech/ColossalAI#3410] [Development | Core]
### Proposal

Current folder structure of zero/gemini is very confusing. The main components are implemented in different folders. E.g.

| Component | Folder |
| :---: | :---: |
| Legacy Zero | zero/ |
| Low level Zero | zero/ |
| Gemini Hook | zero/utils |
| ColoInitContext | utils/model/colo_init_context |
| Chunk/Gemini Manager | gemini/ |
| Zero Optimizer | nn/optimizer/zero_optimizer |
| ZeroDDP/GeminiDDP | nn/parallel |
| zero_wrapper | nn/parallel |

We can reorganize the folder structure. E.g.

```
zero
| --- legacy # all legacy code related to legacy zero
| --- low_level # code related to low level zero optim
       | --- LowLevelZeroOptimizer
| --- gemini 
       | --- GeminiDDP
       | --- GeminiOptimizer
       | --- ColoInitContext
```
### Roadmap
- [x] Reorganize zero/gemini folder structure
- [x] Update examples/applications
- [x] Reorgnaize test folder structure

### Self-service

- [X] I'd be willing to do some initial work on this proposal myself.

**Comment by FrankLeeeee** [comment]
Can you convert this to discussion?

**Comment by FrankLeeeee** [comment]
Looks good to me.

**Comment by kurisusnowdeng** [comment]
I prefer naming the module after Gemini, while zero & chunk management are both its sub-functionalities

**Comment by SauravMaheshkar** [comment]
If I'm not wrong this proposal, has been worked through ?
